airbyte-workers/src/test/java/io/airbyte/workers/temporal/sync/PersistStateActivityTest.java

```diff
@@ -57,4 +67,46 @@ public void testPersist() throws IOException {
   }
 
+  @Test
+  public void testPersistWithInvalidStateDuringMigration() throws IOException {
+    final AirbyteStateMessage stateMessage1 = new AirbyteStateMessage()
+        .withType(AirbyteStateType.STREAM)
+        .withStream(
+            new AirbyteStreamState().withStreamDescriptor(new StreamDescriptor().withName("a").withNamespace("a1"))
+                .withStreamState(Jsons.emptyObject()));
+    final JsonNode jsonState = Jsons.jsonNode(List.of(stateMessage1));
+    final State state = new State().withState(jsonState);
+
+    final ConfiguredAirbyteStream stream = new ConfiguredAirbyteStream().withStream(new AirbyteStream().withName("a").withNamespace("a1"));
+    final ConfiguredAirbyteStream stream2 = new ConfiguredAirbyteStream().withStream(new AirbyteStream().withName("b"));
+    final ConfiguredAirbyteCatalog migrationConfiguredCatalog = new ConfiguredAirbyteCatalog().withStreams(List.of(stream, stream2));
+    final StandardSyncOutput syncOutput = new StandardSyncOutput().withState(state).withOutputCatalog(migrationConfiguredCatalog);
+    Mockito.when(featureFlags.useStreamCapableState()).thenReturn(true);
+    Mockito.when(statePersistence.isMigration(CONNECTION_ID, StateType.STREAM)).thenReturn(true);
+    Assertions.assertThrows(IllegalStateException.class, () -> persistStateActivity.persist(CONNECTION_ID, syncOutput));
+  }
+
+  @Test
+  public void testPersistWithValidStateDuringMigration() throws IOException {
+    final AirbyteStateMessage stateMessage1 = new AirbyteStateMessage()
+        .withType(AirbyteStateType.STREAM)
+        .withStream(
+            new AirbyteStreamState().withStreamDescriptor(new StreamDescriptor().withName("a").withNamespace("a1"))
+                .withStreamState(Jsons.emptyObject()));
+    final AirbyteStateMessage stateMessage2 = new AirbyteStateMessage()
+        .withType(AirbyteStateType.STREAM)
+        .withStream(
+            new AirbyteStreamState().withStreamDescriptor(new StreamDescriptor().withName("b")).withStreamState(Jsons.emptyObject()));
+    final JsonNode jsonState = Jsons.jsonNode(List.of(stateMessage1, stateMessage2));
+    final State state = new State().withState(jsonState);
+
+    final ConfiguredAirbyteStream stream = new ConfiguredAirbyteStream().withStream(new AirbyteStream().withName("a").withNamespace("a1"));
+    final ConfiguredAirbyteStream stream2 = new ConfiguredAirbyteStream().withStream(new AirbyteStream().withName("b"));
+    final ConfiguredAirbyteCatalog migrationConfiguredCatalog = new ConfiguredAirbyteCatalog().withStreams(List.of(stream, stream2));
+    final StandardSyncOutput syncOutput = new StandardSyncOutput().withState(state).withOutputCatalog(migrationConfiguredCatalog);
+    Mockito.when(featureFlags.useStreamCapableState()).thenReturn(true);
+    Mockito.when(statePersistence.isMigration(CONNECTION_ID, StateType.STREAM)).thenReturn(true);
+    persistStateActivity.persist(CONNECTION_ID, syncOutput);
```

Comment:
nit: add an verification to the end of this test that `statePersistence.updatedOrCreateState()` was called (like at the end of the `testPersist()` test)